### PR TITLE
Sprint3

### PR DESCRIPTION
**Pull request by tbjoern**
the linter still complains about table_scan.cpp:205, however seperating initializer lists with a space looks just wrong imo.

### CODE REVIEW

File: src/lib/operators/table_scan.cpp [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+      return std::less<T>();
+    case Type::OpLessThanEquals:
+      return std::less_equal<T>();
+    case Type::OpGreaterThan:
+      return std::greater<T>();
+    case Type::OpGreaterThanEquals:
+      return std::greater_equal<T>();
+  }
+  DebugAssert(false, "Operator not Implemented");
+  return [](T t1, T t2) { return false; };
```
Comment (stephde) [inline src/lib/operators/table_scan.cpp:35]: you might want to throw an exception here instead, since this should never happen

File: src/lib/operators/table_scan.cpp [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+  }
+
+  const auto t_comparator = make_comparator<T>(scan_type);
+  const auto t_value = type_cast<T>(value);
+
+  if (table->chunk_count() == 1) {
+    const auto& chunk = table->get_chunk(ChunkID{0});
+    auto reference_column = std::dynamic_pointer_cast<const ReferenceColumn>(chunk.get_column(column_id));
+    if (reference_column) {
+      auto ref_table = reference_column->referenced_table();
```
Comment (stephde) [inline src/lib/operators/table_scan.cpp:75]: should this reference be declared as `auto&` ?

File: src/lib/operators/table_scan.cpp [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+        if (t_comparator(values[chunk_offset], t_value)) {
+          pos_list->push_back(RowID{chunk_id, chunk_offset});
+        }
+      }
+      continue;
+    }
+
+    auto dict_column = std::dynamic_pointer_cast<const DictionaryColumn<T>>(column);
+    DebugAssert(dict_column,
+                "Unknown Column Type in Table Scan");  // make sure the above is the case; only in debug mode
```
Comment (stephde) [inline src/lib/operators/table_scan.cpp:144]: this comment is not really necessary

File: src/lib/operators/table_scan.cpp [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+        add_all_values(*pos_list, chunk_id, dict_column->size());
+        continue;
+      }
+      if (scan_type == ScanType::OpGreaterThan || scan_type == ScanType::OpGreaterThanEquals) {
+        // every value is smaller than the search value
+        continue;
+      }
+    }
+
+    if (comp_value != ValueID{0} && comp_value != INVALID_VALUE_ID &&
```
Comment (stephde) [inline src/lib/operators/table_scan.cpp:205]: The first two conditions are not really necessary, because if they were true, `continue;` would already have been called above.

File: src/test/operators/print_test.cpp [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+#include "gtest/gtest.h"
+
+#include "operators/get_table.hpp"
+#include "operators/print.hpp"
+#include "storage/storage_manager.hpp"
+#include "storage/table.hpp"
+
+namespace opossum {
+
+// class OperatorsPrintTest : public BaseTest {
```
Comment (stephde) [inline src/test/operators/print_test.cpp:15]: You should probably enable these tests...

File: CMakeLists.txt [hunk 6]
```diff
@@ -25,6 +25,9 @@ set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}) # Put binaries into root
 set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib) # Put libraries into their own dir
 set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -march=native")
 
+find_package( Boost REQUIRED COMPONENTS program_options regex )
+include_directories( ${Boost_INCLUDE_DIRS} )
+
```
Comment (mrks) [inline CMakeLists.txt:30]: How is that related to this Sprint?

File: src/lib/operators/table_scan.cpp [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+  // this holds the positions of all found values
+  auto pos_list = std::make_shared<PosList>();
+  // table that contains the chunk with reference columns
+  auto output_table = std::make_shared<Table>();
+
+  for (ColumnID i{0}; i < table->col_count(); ++i) {
+    output_table->add_column_definition(table->column_name(i), table->column_type(i));
+  }
+
+  const auto t_comparator = make_comparator<T>(scan_type);
```
Comment (mrks) [inline src/lib/operators/table_scan.cpp:68]: What does `t_` mean here?

File: src/lib/operators/table_scan.cpp [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+  const auto t_value = type_cast<T>(value);
+
+  if (table->chunk_count() == 1) {
```
Comment (mrks) [inline src/lib/operators/table_scan.cpp:71]: Why do you have a special case for one chunk? This should be documented (and might be split up in two methods).

File: src/lib/operators/table_scan.cpp [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+    const auto& chunk = table->get_chunk(ChunkID{0});
+    auto reference_column = std::dynamic_pointer_cast<const ReferenceColumn>(chunk.get_column(column_id));
+    if (reference_column) {
```
Comment (mrks) [inline src/lib/operators/table_scan.cpp:74]: This looks like you are assuming all temporary tables to have exactly one chunk. However, the table scan of a different group could produce more than one chunk as a result.

File: src/lib/operators/table_scan.cpp [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+      auto ref_table = reference_column->referenced_table();
+      auto ref_column_id = reference_column->referenced_column_id();
+      auto ref_pos_list = reference_column->pos_list();
+
+      auto ref_chunk_id = ChunkID{0};
+      const std::vector<T>* ref_val_vector = nullptr;
```
Comment (mrks) [inline src/lib/operators/table_scan.cpp:80]: We don't want to use C-style pointers. In this case, optional would be semantically better.

File: src/lib/operators/table_scan.cpp [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+      auto ref_dict_col = std::shared_ptr<const DictionaryColumn<T>>();
+      // go through the position list
+      for (const RowID& r : *ref_pos_list) {
+        // we assume that the RowIDs are sorted by ChunkID (ascending)
+        // thus we only have to fetch the columns once per referenced chunk
```
Comment (mrks) [inline src/lib/operators/table_scan.cpp:85]: Where are you taking that assumption from?

File: src/lib/operators/table_scan.cpp [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+        if (ref_chunk_id != r.chunk_id || (!ref_val_vector && !ref_dict_col)) {
+          ref_chunk_id = r.chunk_id;
```
Comment (mrks) [inline src/lib/operators/table_scan.cpp:87]: This is a good example why abbreviations are almost never worth it.

File: src/lib/operators/table_scan.cpp [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+          const auto& ref_chunk = ref_table->get_chunk(ref_chunk_id);
+          ref_dict_col = std::dynamic_pointer_cast<const DictionaryColumn<T>>(ref_chunk.get_column(ref_column_id));
+          auto ref_val_col = std::dynamic_pointer_cast<const ValueColumn<T>>(ref_chunk.get_column(ref_column_id));
+          ref_val_vector = ref_val_col ? &ref_val_col->values() : nullptr;
+        }
+        if (ref_dict_col) {
+          if (t_comparator(ref_dict_col->get(r.chunk_offset), t_value)) {
```
Comment (mrks) [inline src/lib/operators/table_scan.cpp:94]: You are not using the dictionary optimization here, are you?

File: src/lib/operators/table_scan.cpp [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+          }
+          continue;
+        }
+        if (ref_val_vector) {
+          if (t_comparator((*ref_val_vector)[r.chunk_offset], t_value)) {
+            pos_list->push_back(r);
+          }
+          continue;
+        }
+        DebugAssert(false, "Unexpected Column Type when dereferencing ReferenceColumn");
```
Comment (mrks) [inline src/lib/operators/table_scan.cpp:105]: `Fail`?

File: src/lib/operators/table_scan.cpp [hunk 15]
```diff
@@ -0,0 +1,241 @@
...
+        reference_chunk.add_column(reference_column);
+      }
+
+      output_table->emplace_chunk(std::move(reference_chunk));
+
+      return output_table;
+    }
+  }
+
+  auto ValueID_comparator = make_comparator<ValueID>(scan_type);
```
Comment (mrks) [inline src/lib/operators/table_scan.cpp:121]: Naming

File: src/lib/operators/table_scan.cpp [hunk 16]
```diff
@@ -0,0 +1,241 @@
...
+      case ScanType::OpLessThanEquals:
+      case ScanType::OpGreaterThan:
+        comp_value = dict_column->upper_bound(t_value);
+        break;
+    }
+
+    if (comp_value == ValueID{0u}) {
+      if (scan_type == ScanType::OpLessThanEquals || scan_type == ScanType::OpLessThan) {
+        // value is smaller than all the values in the column, there are no matches
+        // we can go to the next chunk
```
Comment (mrks) [inline src/lib/operators/table_scan.cpp:182]: Shouldn't `OpLessThanEquals `work with value id 0?

File: src/lib/operators/table_scan.cpp (hunk 4 above)
Comment (mrks) [inline src/lib/operators/table_scan.cpp:205]: This could use a comment.

File: src/lib/operators/table_scan.cpp [hunk 17]
```diff
@@ -0,0 +1,241 @@
...
+    reference_chunk.add_column(reference_column);
+  }
+
+  output_table->emplace_chunk(std::move(reference_chunk));
+
+  return output_table;
+}
+
+void add_all_values(PosList& pos_list, ChunkID chunk_id, ChunkOffset number_of_values) {
+  pos_list.reserve(number_of_values);
```
Comment (mrks) [inline src/lib/operators/table_scan.cpp:233]: This is potentially called for multiple chunks, isn't it? In that case, you are not reserving enough.

File: src/lib/operators/table_scan.cpp [hunk 18]
```diff
@@ -0,0 +1,241 @@
...
+    switch (scan_type) {
+      case ScanType::OpEquals:
+      case ScanType::OpNotEquals:
+      case ScanType::OpLessThan:
+      case ScanType::OpGreaterThanEquals:
+        comp_value = dict_column->lower_bound(t_value);
+        break;
+      case ScanType::OpLessThanEquals:
+      case ScanType::OpGreaterThan:
+        comp_value = dict_column->upper_bound(t_value);
```
Comment (rs22) [inline src/lib/operators/table_scan.cpp:175]: I think using upper_bound to find the comp_value might be a logical problem (unless I'm missing some further transformation here): It *never* returns the position of t_value in the dictionary, but the position of the next-bigger element (if any). So I guess you're comparing the wrong value later on.
http://en.cppreference.com/w/cpp/algorithm/upper_bound

File: src/lib/operators/table_scan.cpp (hunk 4 above)
Comment (rs22) [reply src/lib/operators/table_scan.cpp:205]: Oh, I see this is where you fix the comp_value when using upper_bound. A little confusing.

It's not quite obvious how the control flow works here for different operator/value-edge-cases. 
It might introduce some duplicated code but I think it could help to introduce separate methods handling only one of the operators at a time.

File: src/lib/operators/table_scan.cpp [hunk 19]
```diff
@@ -0,0 +1,241 @@
...
+    }
+
+    auto attribute_vector = dict_column->attribute_vector();
+
+    for (ChunkOffset i = 0; i < attribute_vector->size(); ++i) {
+      if (ValueID_comparator(attribute_vector->get(i), comp_value)) {
+        pos_list->push_back(RowID{chunk_id, i});
+      }
+    }
+  }  // end for (chunks)
```
Comment (rs22) [inline src/lib/operators/table_scan.cpp:217]: I think it could really help to invent some methods in order to split up not only this for-loop but also the entire method. 

For example, the handling of reference columns (under if (chunk_count == 1)) would have been a good candidate.

File: src/lib/operators/table_scan.cpp [hunk 20]
```diff
@@ -0,0 +1,241 @@
...
+
+  output_table->emplace_chunk(std::move(reference_chunk));
+
+  return output_table;
+}
+
+void add_all_values(PosList& pos_list, ChunkID chunk_id, ChunkOffset number_of_values) {
+  pos_list.reserve(number_of_values);
+  for (ChunkOffset i = 0; i < number_of_values; i++) {
+    pos_list.push_back(RowID{chunk_id, i});
```
Comment (rs22) [inline src/lib/operators/table_scan.cpp:235]: You could've also used `emplace_back`, now that you've already reserved the neccesary space.
